mlton 20130715 (new formula)

MatthewFluet opened the pull request:
MLton is whole-program, optimizing compiler for Standard ML.

A previous mlton formula simply installed the upstream binary release
and was moved to the boneyard as a binary-only formula (see
Homebrew/homebrew#21780).

This new mlton formula builds from source, using the upstream binary
release to bootstrap; this seems to be an acceptable approach for
self-hosting compilers (e.g., the ghc formula).

Closes Homebrew/homebrew#48150.

mikemcquaid commented on Library/Formula/mlton.rb:
```diff
@@ -0,0 +1,131 @@
+
+# MLton is a self-hosting compiler for Standard ML.
+# In order to build from source, this formula uses the corresponding
+# upstream binary release to bootstrap.
```
Can remove these comments

mikemcquaid commented on Library/Formula/mlton.rb:
```diff
@@ -0,0 +1,131 @@
+  # The upstream source release hard codes common locations for GMP;
+  # on darwin, these include '/opt/local' (MacPorts) and '/sw' (Fink).
+  # Replace these with placeholders for the Homebrew GMP location.
+  patch :DATA
```
Can you please submit this patch to the upstream developers of this project and add a link to the upstream patch submission and explanation of why the patch is needed in a comment in the formula file. Thanks!

MatthewFluet replied:
The source package has no `configure` step and the upstream policy has been to include common locations for GMP on various target platforms.  The upstream [installation instructions](http://mlton.org/Installation) explicitly mention editing the `bin/mlton` script to find `gmp.h` and `libgmp.a`.  This patch wouldn't be appropriate for upstream, since it removes support for Fink and MacPorts.  Support for default Homebrew prefix is already implicit upstream, since `-I/usr/local/include` and `-L/usr/local/lib` are used; patching upstream with a placeholder for a non-standard Homebrew prefix seems heavy weight.

I could remove the patch and achieve the same effect with only two `inreplace` lines, at the expense of leaving `/opt/local/{include,lib}` and `/sw/{include,lib}` in the search paths.

apjanke replied:
Requiring the user to manually edit the script to point to the installation location sounds like a bug worth reporting upstream. That info (`PREFIX`) is available when `make` is called, and the change could be done automatically as part of the `script` target in the `Makefile`, where it produces the `mlton` script.

```
script:
	$(CP) bin/mlton-script "$(MLTON)"
	chmod a+x "$(MLTON)"
	$(CP) "$(SRC)/bin/platform" "$(LIB)"
...
```

In fact, it seems like all or most of the hardcoded `/usr/local` paths in `mlton-script` (and maybe `bin/add-cross` and `bin/regression` too) should be replaced with `PREFIX` at `make` time.

The GMP location, if it's allowed to vary from `PREFIX`, could be parameterized as a Makefile variable, which could preserve the existing behavior and let Homebrew or other custom installations repoint it using a `make` argument, instead of requiring a patch.

Regardless, if we're repointing via either patches or a `make` variable, the Homebrew locations should maybe be pointing to `brew --prefix` instead of `Formula["gmp"].opt_prefix`. The `gmp` formula is not keg-only, so its `opt_prefix` isn't specifically required – it'll be linked in to `$(brew --prefix)/lib` and so on. And I think you'd want to have the other libraries installed in this prefix available too. That would be more like the default MLton behavior, which pulls in everything in `/usr/local`.

(If you end up submitting upstream, you might also ask about this section in the Makefile.
```
PREFIX := /usr
ifeq ($(findstring $(TARGET_OS), darwin freebsd solaris), $(TARGET_OS))
PREFIX := /usr/local
endif
```
It seems odd to make the `/usr` vs `/usr/local` distinction on the basis of OS version; seems like it should be `/usr/local` everywhere that doesn't have a special platform-dependent location like `/mingw`.)

mikemcquaid replied:
It's not so much about the patch vs inreplace but instead this being something upstream should support in the medium-term so we need it to be reported to them as both `inreplace`s and patches have a habit of breaking in future releases.

mikemcquaid commented on Library/Formula/mlton.rb:
```diff
@@ -0,0 +1,131 @@
+    chmod "a+x", bootstrap_bin/"mlton"
+    ENV.prepend_path "PATH", bootstrap_bin
+
+    ENV.deparallelize
```
Can you please submit these `inreplace` changes and the parallelisation issue to the upstream developers of this project and add a link to the upstream patch submission and explanation of why the patch is needed in a comment in the formula file. Thanks!

MatthewFluet replied:
Regarding the `inreplace` changes: I'm not sure what would be submitted upstream.  This is patching the upstream binary release, which assumes that it will be unpacked in `/` (yielding, `/usr/local/lib/mlton` for the support files).  Again, the [install instructions](http://mlton.org/Installation) explicitly mention editing the `lib` variable in the `mlton` script if the binary package is installed somewhere else.

Regarding the deparallelize: I can certainly open an upstream issue.  However, a previous attempt at making the MLton build parallel safe determined that the patch would be quite invasive; see [this message](https://sourceforge.net/p/mlton/mailman/message/31031131/).

apjanke replied:
>  This is patching the upstream binary release, which assumes that it will be unpacked in /

You could report that assumption and the need for manually editing the script as a bug. If it really is relocatable, like we're using here, then the `lib` variable could be set based on `$0` to detect the script's location, like `bin/mlton-script` does in the source distribution.

```
dir=`dirname "$0"`
lib=`cd "$dir/../lib" && pwd`
```

Looks like it might need a symlink-dereferencing step, though.

WRT deparallelize: a link to that message in a comment should be sufficient. Or, if they're actively using the GitHub issue tracker now, it would be nice to have it as an issue there, so it would pick up future comments and status updates.

apjanke commented on Library/Formula/mlton.rb:
```diff
@@ -0,0 +1,131 @@
+-if [ -d '/sw/lib' ]; then
+-        darwinLinkOpts="$darwinLinkOpts -L/sw/lib"
+-fi
++gmpHomebrewCCOpts=''
```
If we are going to patch it like this, I think we should probably leave all the existing `$darwinLinkOpts` code in place, and just add another line to get the Homebrew prefix (determined at install time) in there and in a `$darwinCcOpts`. That will minimize the difference in behavior between the Homebrew-installed and regular MLtons.

apjanke commented:
With respect to all the patches & comments so far (except the `deparallelize`), I think this comes down to MLton's installer's support for `prefix` being incomplete. That's pretty basic, and I think it's reasonable to ask upstream to improve it, instead of working around it in the Homebrew formula. That way there's less maintenance burden on Homebrew, it's less likely to break, and all MLton users get the benefit of a working `prefix`. Might allow MLton to simplify and remove its MacPorts/Fink/Homebrew special-casing, too.

MatthewFluet commented:
Thanks for the feedback.  I readily acknowledge that MLton's build system currently leaves much to be desired; I hope to have a student addressing some of these issues this semester.  I'll create some issues at https://github.com/MLton/mlton and consider if there are some simple patches that can be applied upstream that would minimize the necessary Homebrew customization.

@apjanke Regarding using `Formula["gmp"].opt_prefix`: It seemed to occur regularly with formulas that had an explicit `--with-gmp=...` configuration step, especially other compilers (e.g, `gcc.rb`, `guile.rb`, `gnu-cobol.rb`).  Of course, many formulas that depend on GMP don't explicitly configure the GMP location; presumably, they pick up a non-standard Homebrew prefix via CFLAGS?

apjanke commented:
> Regarding using Formula["gmp"].opt_prefix: It seemed to occur regularly with formulas...

I'm a relatively new maintainer, so I might not have it straight on this one. Doing the `Formula["foo"].opt_prefix` is for when you want to pick up one specific `foo` library, regardless of sandboxing or environment sanitization. Picking up `$(brew --prefix)` is when you want to link/include all the Homebrew-managed libraries. (I *think*; someone correct me if I'm wrong here.) I'm not sure which is most appropriate for MLton, but `$(brew --prefix)` matches its non-brewed behavior, where it's picking up everything in MacPorts/Fink/other package managers, so that seemed more appropriate. Since MLton is a compiler which might have other (C) library dependencies not known at install time, it seemed like the broader one might be the right one, especially in terms of matching run time behavior for the non-brewed installations.

For those `-L`/`-I` inclusions, are you looking to pick up *just* GMP, or is this a broader setting for the compiler where you might be picking up arbitrary other libraries?

apjanke commented:
As far as the build system `prefix` support, it seems to me like a couple simple changes to the Makefile along with some basic text replacement in the `bin/` wrapper scripts might be all you need. If you'd entertain a PR from an outsider/ML newbie with some Makefile experience, I'd be happy to submit a couple patch suggestions.

apjanke commented:
> ... presumably, they pick up a non-standard Homebrew prefix via CFLAGS?

...and, yeah, I got off track, but (from my understanding), tools pick up the nonstandard Homebrew prefix via `CFLAGS` and other environment variables, or `make` arguments (generally supplied by formulae). And to have the non-standard prefix they're installed to "baked in" to their default paths as part of the installation process (which is a normal part of the installation process for packages that support `prefix`). Compilers and build tools used within Homebrew get those environment variables and `make` arguments supplied via other Homebrew formulae.

I.e. it's a package's job to respect `prefix` for both targeting and discovery, and the package manager's job to supply other paths. (With some middle ground of auto-discovery of well-known locations.) (Other folks feel free to chime in, since I'm making broad normative statements here.) That is, it's `mlton`'s job to provide hooks for relocating the `mlton` installation and pointing to the GMP library, but it shouldn't have to make efforts to discover the package managers' location(s) for libraries or whether particular package managers are installed. That's the package managers' job.

mikemcquaid commented:
> With respect to all the patches & comments so far (except the deparallelize), I think this comes down to MLton's installer's support for prefix being incomplete. That's pretty basic, and I think it's reasonable to ask upstream to improve it, instead of working around it in the Homebrew formula.

I agree here; I'd rather wait until this is a lot simpler and and works without patches before we include it.

MatthewFluet commented:
Thanks again for all of the feedback.  I've created MLton/mlton#131 (and a number of linked issues) to capture the main suggestions from this PR discussion.

@apjanke I'd certainly be interested in reviewing your specific suggestions for the `Makefile`s.  To quickly answer your question about `-I`/`-L` inclusions in the `bin/mlton` script, they need only include GMP (but on most platforms, the location of GMP probably also has lots of other stuff and is probably already on the default include/link paths).  The `-link-opt` option is a documented option for users to specify additional items at the linking step, usually `-L...` and `-l...` options.  The idea is that a standard SML program (i.e., one using anything from the SML Basis Library but not using any FFI extensions) should compile without any additional user options; this only requires libc and GMP.  But, if the SML program uses the FFI extensions to import C functions from a library, then the user may be required to give link options (just as they would with a C program, if the library were not on the default paths).

@mikemcquaid Regarding "works without patches" and a future mlton 20130715 formula: would a `patch` with `url "https://github.com/MLton/mlton/commit/deadbeef.diff"` (for a commit merged into `master`) be acceptable?  Since the (top-level) `Makefile` and `bin/mlton-script` haven't really changed, I'm hopeful that the necessary changes would apply cleanly to both the mlton 20130715 source release and `master`.

mikemcquaid commented:
> @mikemcquaid Regarding "works without patches" and a future mlton 20130715 formula: would a patch with url "https://github.com/MLton/mlton/commit/deadbeef.diff" (for a commit merged into master) be acceptable? Since the (top-level) Makefile and bin/mlton-script haven't really changed, I'm hopeful that the necessary changes would apply cleanly to both the mlton 20130715 source release and master.

@MatthewFluet Yeh, that's be fine; I basically want to just see that these changes have been submitted and accepted upstream.

MatthewFluet commented:
I presume the coveralls failure is due to now being far behind `master`.  And the travis failure (`OSX=10.9 HOMEBREW_RUBY=1.8.7`) seems unrelated to the mlton formula.

MatthewFluet commented:
@apjanke @mikemcquaid I'd like to know how I should proceed with this pull request.  As far as I can tell, the test failures from the last commit are not directly related to the mlton formula.

apjanke commented:
Upstream has [merged that patch in head](https://github.com/MLton/mlton/pull/136) so I think we're good to go here.

Yeah, the Travis CI failure is unrelated (issues with the test machine), and coveralls may well be due to being behind master. At this point, the Jenkins build has aged out, which I think we need for the bottle. If you squash this PR to a single commit (which we prefer for changes) and rebase it on `master`, that will trigger another Jenkins build and get a fresh Coveralls report, and we should be ready to merge at that point.

MatthewFluet commented:
Hmm... I don't know why the Travis and Coveralls checks didn't trigger.

apjanke commented:
We've turned off Travis CI checks for PRs for the moment, so that's expected. Maybe Coveralls depends on Travis?

At any rate, Jenkins is the critical test, and that passed.

Merged. Thank you for you contribution to Homebrew, and your patience on this!
